### ISSUE TITLE
mmlu evaluation not working

### ISSUE THREAD

**Issue by zhengyang-wang** [facebookresearch/lingua#28]
Thanks for open-sourcing this! I tested the repo with a small job and found that the 0-shot MMLU score remains 0.22945449366187154 throughout the training. I checked the dumped evaluation details and found that the negative log-likelihood is always 0. For example,
```
{'doc_id': 0,
 'doc': {'question': " Just war theory's principle of military necessity belongs to",
  'subject': 'moral_disputes',
  'choices': ['jus in bello.',
   'jus ad bellum.',
   'moral nihilism.',
   'all of the above'],
  'answer': 0},
 'target': 0,
 'arguments': [["The following are multiple choice questions (with answers) about moral disputes.\n\nJust war theory's principle of military necessity belongs to\nA. jus in bello.\nB. jus ad bellum.\nC. moral nihilism.\nD. all of the above\nAnswer:",
   ' A'],
  ["The following are multiple choice questions (with answers) about moral disputes.\n\nJust war theory's principle of military necessity belongs to\nA. jus in bello.\nB. jus ad bellum.\nC. moral nihilism.\nD. all of the above\nAnswer:",
   ' B'],
  ["The following are multiple choice questions (with answers) about moral disputes.\n\nJust war theory's principle of military necessity belongs to\nA. jus in bello.\nB. jus ad bellum.\nC. moral nihilism.\nD. all of the above\nAnswer:",
   ' C'],
  ["The following are multiple choice questions (with answers) about moral disputes.\n\nJust war theory's principle of military necessity belongs to\nA. jus in bello.\nB. jus ad bellum.\nC. moral nihilism.\nD. all of the above\nAnswer:",
   ' D']],
 'resps': [[[0.0, True]], [[0.0, True]], [[0.0, True]], [[0.0, True]]],
 'filtered_resps': [[0.0, True], [0.0, True], [0.0, True], [0.0, True]],
 'doc_hash': 'bb0de79f79411c47783968714ec9fe3c69d89753e22c88f044420a7e00049a15',
 'prompt_hash': 'c0f269e09cf44177328b16fb43734d6675416710a24286fea49dfad5663d2fb4',
 'target_hash': '5feceb66ffc86f38d952786c6d696c79c2dbc239dd4e91b46729d73a27fb57e9',
 'acc': 1.0}
```

The other benchmarks seem normal, e.g. (piqa)
```
{'doc_id': 0,
 'doc': {'goal': "How do I ready a guinea pig cage for it's new occupants?",
  'sol1': 'Provide the guinea pig with a cage full of a few inches of bedding made of ripped paper strips, you will also need to supply it with a water bottle and a food dish.',
  'sol2': 'Provide the guinea pig with a cage full of a few inches of bedding made of ripped jeans material, you will also need to supply it with a water bottle and a food dish.',
  'label': 0},
 'target': 0,
 'arguments': [["Question: How do I ready a guinea pig cage for it's new occupants?\nAnswer:",
   ' Provide the guinea pig with a cage full of a few inches of bedding made of ripped paper strips, you will also need to supply it with a water bottle and a food dish.'],
  ["Question: How do I ready a guinea pig cage for it's new occupants?\nAnswer:",
   ' Provide the guinea pig with a cage full of a few inches of bedding made of ripped jeans material, you will also need to supply it with a water bottle and a food dish.']],
 'resps': [[[-120.0, False]], [[-122.5, False]]],
 'filtered_resps': [[-120.0, False], [-122.5, False]],
 'doc_hash': 'ab177c9b9ad0fd48149e873e3d4804752991338a90c2072f52b975f86a7ca78e',
 'prompt_hash': '14e2e90bdc64add59c76a88c2efb217a132b459947f9f2e3bbe8580b71beb533',
 'target_hash': '5feceb66ffc86f38d952786c6d696c79c2dbc239dd4e91b46729d73a27fb57e9',
 'acc': 1.0,
 'acc_norm': 1.0}
```

The saved eval config is 
```
harness:
  tasks:
  - hellaswag
  - nq_open
  - piqa
  - winogrande
  - arc
  - race
  - mmlu
  num_fewshot: null
  device: null
  use_cache: null
  cache_requests: false
  rewrite_requests_cache: false
  delete_requests_cache: false
  limit: null
  bootstrap_iters: 100000
  check_integrity: false
  write_out: false
  log_samples: true
  system_instruction: null
  apply_chat_template: false
  fewshot_as_multiturn: false
  gen_kwargs: null
  verbosity: INFO
  predict_only: false
  random_seed: 0
  numpy_random_seed: 1234
  torch_random_seed: 1234
  fewshot_random_seed: 1234
wandb: null
global_step: 60000
```

It might be because MMLU's answer is only one token and the eval script misses the last token? Looking forward to fixes.

**Comment by BadrYoubiIdrissi** [comment]
Hey thank you for the issue ! Let me try to reproduce that and see what is the issue. We've only recently switched to lm-harness from our internal evals so some things might be broken, sorry about that!

**Comment by tanishqkumar** [comment]
+1

**Comment by ryoungj** [comment]
Same issue occurred on arc and lambada evals.

I think the bug comes from here: https://github.com/facebookresearch/lingua/blob/main/apps/main/eval.py#L134-L136
```
for p, ll, gr in zip(prompts, lls, greedy):
    p_len = len(self.generator.tokenizer.encode(p, add_bos=True, add_eos=False))
    results.append((ll[p_len:].sum().item(), gr[p_len:].all().item()))
```

The `p_len` corresponds to the token length of the full prompt, while both `ll` and `gr` are of length `p_len-1` because the last token prediction is excluded. As a result, both `ll[p_len:]` and `gr[p_len:]` become an empty list when there is only one single token. 

A simple fix that applies `p_len -= 1` gives the correct result. Happy to create PR if it is reasonable

**Comment by zhengyang-wang** [comment]
Thanks @ryoungj . Would you elaborate a bit on why it happens when there is only one single output token?
From your current explanation, it seems that it is always one-token off, so that other metrics are also affected, but just less obvious from results.

**Comment by ryoungj** [comment]
I edited my previous explanation to be more clear: the key is that the indexing of the continuation likelihoods is 1 off due to the BOS token.

Here is an example:
```
Example:
- Prompt is like "p_1 p_2 p_3", continuation is like "a_1 a_2"
- Inputs is like "p_1 p_2 p_3 a_1 a_2"
- p_len is len(["bos", "p_1", "p_2", "p_3"]) -> 4 = len(prompt) + 1
- both ll and greedy are like [l_{p_1}, l_{p_2}, l_{p_3}, l_{a_1}, l_{a_2}] -> 5 = len(inputs) = len(prompt) + len(continuation)
- We need to compute the log likelihood of "a_1 a_2", which should be start from index 3 = len(prompt) = p_len - 1
```

**Comment by zhengyang-wang** [comment]
Thanks a lot for your explanation!

So we can also fix it by changing
`p_len = len(self.generator.tokenizer.encode(p, add_bos=True, add_eos=False))`
to
`p_len = len(self.generator.tokenizer.encode(p, add_bos=False, add_eos=False))`?

**Comment by akhauriyash** [comment]
There was some plan to migrate the evaluation suite away from lm-evaluation-harness, was there any progress on this end?

Thanks

**Comment by mathuvu** [comment]
This has been fixed in #47. Thank you for your help!
